### PR TITLE
vpc

### PR DESCRIPTION
**Pull request by narkarum**

### CODE REVIEW

**Comment by arahamad** [comment]
@narkarum , As discussed can you please cleanup code which are not needed. I code in vpc/provider and vpc/client has un-used code

File: volume-providers/vpc/util/unit_util.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+// DiskConfig ...
+type DiskConfig struct {
+	Count    int
+	Size     int // Size is always in GB
+	DiskType string
+}
+
+// GetMemorySizeInGB processes a string, and extracts the memory size. If the provided
+// unit is in MB, this is converted to GB
+func GetMemorySizeInGB(value string) (memoryGB int, err error) {
```
Comment (arahamad) [inline volume-providers/vpc/util/unit_util.go]: I don't think so its useful for volume

File: volume-providers/vpc/util/unit_util.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+import (
+	"github.com/IBM/ibmcloud-storage-volume-lib/lib/utils/reasoncode"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// Constants
+const (
+	DiskUnitsMB string = "MB"
```
Comment (arahamad) [inline volume-providers/vpc/util/unit_util.go]: Please check these const values are useful or not, I don't think so its useful for volume

File: volume-providers/vpc/util/unit_util.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+		memoryGB = memory
+	}
+
+	return
+
+}
+
+// GetNetworkPortSpeedInMbps processes a string, and extracts the network speed. If the provided
+// speed is in Gbpg, this is converted to Mbps
+func GetNetworkPortSpeedInMbps(value string) (networkSpeedMbps int, err error) {
```
Comment (arahamad) [inline volume-providers/vpc/util/unit_util.go]: I don't think so its useful for volume

File: volume-providers/vpc/util/unit_util.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+//   <count1>x_<gpu_type1>;<count2>x_<gpu_type2>...
+//
+// It returns an array of SL keys with one entry for each instance of
+// a GPU represented by the spec.
+//
+// e.g. 1xK80;2xV100 will produce an array with 3 entries:
+// [0] = K80
+// [1] = V100
+// [2] = V100
+func ProcessGPUConfig(configuration string, permittedGPUTypes []string) (gpuKeys []string, err error) {
```
Comment (arahamad) [inline volume-providers/vpc/util/unit_util.go]: I don't think so its useful for volume, applicable for other method of this file

File: volume-providers/vpc/util/unit_util_test.go [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+ * IBM Confidential
+ * OCO Source Materials
+ * IBM Cloud Container Service, 5737-D43
+ * (C) Copyright IBM Corp. 2018 All Rights Reserved.
+ * The source code for this program is not  published or otherwise divested of
+ * its trade secrets, irrespective of what has been deposited with
+ * the U.S. Copyright Office.
+ ******************************************************************************/
+
+package util
```
Comment (arahamad) [inline volume-providers/vpc/util/unit_util_test.go]: Please update this file as per actual functions in the imp file

File: lib/provider/constants.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * IBM Confidential
+ * OCO Source Materials
+ * IBM Cloud Container Service, 5737-D43
+ * (C) Copyright IBM Corp. 2018 All Rights Reserved.
+ * The source code for this program is not  published or otherwise divested of
+ * its trade secrets, irrespective of what has been deposited with
+ * the U.S. Copyright Office.
+ ******************************************************************************/
+
+package provider
```
Comment (arahamad) [inline lib/provider/constants.go]: we already have this file at `lib/provider/constants.go` . can we use that

File: lib/provider/vpc_data_types.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ * The source code for this program is not  published or otherwise divested of
+ * its trade secrets, irrespective of what has been deposited with
+ * the U.S. Copyright Office.
+ ******************************************************************************/
+
+package provider
+
+// VPC specific	parameters
+type VPCVolume struct {
+	Href              string              `json:"href,omitempty"`
```
Comment (arahamad) [inline lib/provider/vpc_data_types.go]: I don't think `Href` is useful for the user, can we keep only variable which are useful for user as this is user(client) facing struct.

same with CRN, Generation and Profile

File: lib/utils/reasoncode/vpc_messages_en.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
+/*******************************************************************************
+ * IBM Confidential
+ * OCO Source Materials
+ * IBM Bluemix Container Registry, 5737-D42
+ * (C) Copyright IBM Corp. 2018 All Rights Reserved.
+ * The source code for this program is not  published or otherwise divested of
+ * its trade secrets,  * irrespective of what has been deposited with
+ * the U.S. Copyright Office.
+ ******************************************************************************/
+package reasoncode
```
Comment (arahamad) [inline lib/utils/reasoncode/vpc_messages_en.go]: Can we move this message file to implementation folder i.e vpc provider not on interface level

File: lib/utils/reasoncode/vpc_messages_en.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+var messages_en = map[string]Message{
```
Comment (arahamad) [inline lib/utils/reasoncode/vpc_messages_en.go]: There are many text modifications required but lets get it reviewed by Doc team

File: volume-providers/vpc/client/models/constants.go [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+ * The source code for this program is not  published or otherwise divested of
+ * its trade secrets, irrespective of what has been deposited with
+ * the U.S. Copyright Office.
+ ******************************************************************************/
+
+package models
+
+const (
+	// APIVersion is the target RIaaS API spec version
+	APIVersion = "2019-01-01"
```
Comment (arahamad) [inline volume-providers/vpc/client/models/constants.go]: Is it changeable, if yes can we make it configurable

File: volume-providers/vpc/client/params_test.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ * the U.S. Copyright Office.
+ ******************************************************************************/
+
+package client_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.ibm.com/alchemy-containers/armada-provider-riaas/apiclient/riaas/client"
```
[the remark this answers is not in the document]
Comment (narkarum) [reply volume-providers/vpc/client/params_test.go]: DONE

File: volume-providers/vpc/client/client_test.go [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.ibm.com/alchemy-containers/armada-provider-riaas/apiclient/riaas/client"
```
[the remark this answers is not in the document]
Comment (narkarum) [reply volume-providers/vpc/client/client_test.go]: DONE

File: volume-providers/vpc/client/models/vpc.go [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+ * (C) Copyright IBM Corp. 2018 All Rights Reserved.
+ * The source code for this program is not  published or otherwise divested of
+ * its trade secrets, irrespective of what has been deposited with
+ * the U.S. Copyright Office.
+ ******************************************************************************/
+
+package models
+
+// VPC ...
+type VPC struct {
```
[the remark this answers is not in the document]
Comment (narkarum) [reply volume-providers/vpc/client/models/vpc.go]: DONE

File: volume-providers/vpc/provider/session.go [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+	VolumeType         provider.VolumeType
+	Provider           provider.VolumeProvider
+	Apiclient          riaas.RegionalAPI
+	Logger             *zap.Logger
+}
+
+const (
+	// VPC storage provider
+	VPC            = provider.VolumeProvider("VPC")
+	VPCEndurance   = provider.VolumeProviderType("VPC_ENDURANCE")
```
[the remark this answers is not in the document]
Comment (narkarum) [reply volume-providers/vpc/provider/session.go]: DONE

File: volume-providers/vpc/client/volume/snapshot_service.go [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+package volume
+
+import (
+	"github.com/IBM/ibmcloud-storage-volume-lib/volume-providers/vpc/client"
+	"github.com/IBM/ibmcloud-storage-volume-lib/volume-providers/vpc/client/models"
+)
+
+// Snapshot operations
+type SnapshotManager interface {
```
[the remark this answers is not in the document]
Comment (narkarum) [reply volume-providers/vpc/client/volume/snapshot_service.go]: DONE

File: volume-providers/vpc/util/unit_util.go (hunk 1 above)
Comment (narkarum) [reply volume-providers/vpc/util/unit_util.go]: DONE

File: volume-providers/vpc/util/unit_util.go (hunk 2 above)
Comment (narkarum) [reply volume-providers/vpc/util/unit_util.go]: DONE

File: volume-providers/vpc/util/unit_util.go (hunk 3 above)
Comment (narkarum) [reply volume-providers/vpc/util/unit_util.go]: DONE

File: volume-providers/vpc/util/unit_util.go (hunk 4 above)
Comment (narkarum) [reply volume-providers/vpc/util/unit_util.go]: DONE

File: volume-providers/vpc/util/unit_util_test.go (hunk 5 above)
Comment (narkarum) [reply volume-providers/vpc/util/unit_util_test.go]: DONE

File: lib/provider/vpc_data_types.go (hunk 7 above)
Comment (narkarum) [reply lib/provider/vpc_data_types.go]: NEEDED

File: lib/provider/constants.go (hunk 6 above)
Comment (narkarum) [reply lib/provider/constants.go]: DONE

File: lib/utils/reasoncode/vpc_messages_en.go (hunk 9 above)
Comment (narkarum) [reply lib/utils/reasoncode/vpc_messages_en.go]: MOVED

File: lib/utils/reasoncode/vpc_messages_en.go (hunk 8 above)
Comment (narkarum) [reply lib/utils/reasoncode/vpc_messages_en.go]: DONE

File: volume-providers/vpc/client/riaas/riaas.go [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/IBM/ibmcloud-storage-volume-lib/volume-providers/vpc/client/volume"
+)
+
+// RegionalAPI is the main interface for the RIAAS API client. From here, service
+// objects for the individual parts of the API can be obtained
+//go:generate counterfeiter -o fakes/regional_api.go --fake-name RegionalAPI . RegionalAPI
+type RegionalAPI interface {
+	Login(token string) error
+
+	Volume() volume.VolumeManager
```
[the remark this answers is not in the document]
Comment (narkarum) [reply volume-providers/vpc/client/riaas/riaas.go]: DONE

File: volume-providers/auth/ims_token_test.go [hunk 17]
```diff
@@ -0,0 +1,233 @@
+/*******************************************************************************
```
[the remark this answers is not in the document]
Comment (narkarum) [reply volume-providers/auth/ims_token_test.go]: DONE

File: volume-providers/vpc/provider/provider.go [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/IBM/ibmcloud-storage-volume-lib/volume-providers/vpc/client/riaas"
+	"github.com/rs/xid"
+	"go.uber.org/zap"
+	"net/http"
+	"os"
+	"time"
+)
+
+const (
+	armadaDisplayName      = "IBM Cloud container service"
```
[the remark this answers is not in the document]
Comment (narkarum) [reply volume-providers/vpc/provider/provider.go]: DONE

File: volume-providers/vpc/client/models/status_type.go [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+package models
+
+// StatusType ...
+type StatusType string
+
+// String ...
+func (i StatusType) String() string { return string(i) }
+
+// Status values
+const (
```
[the remark this answers is not in the document]
Comment (arahamad) [reply volume-providers/vpc/client/models/status_type.go]: Done

**Pull request merged**